src/gp_doctest.cpp

```diff
@@ -1989,3 +1987,128 @@ TEST_CASE("Top-Pruning: Likelihoods") {
       fasta_path_six, newick_path_six_simple, false, false);
   CHECK_MESSAGE(test_3, "Six Taxa Multi Tree failed.");
 }
+
+// Runs an instance of TPEngine for two DAGs: DAG_1, a simple DAG, and DAG_2, a DAG
+// formed from DAG_1 plus all of its adjacent NNIs. Both DAGs PVs are populated and
+// their edge TP likelihoods are computed.  Then DAG_1's adjacent proposed NNI
+// likelihoods are computed using only PVs from the pre-NNI already contained in DAG_1.
+// Finally, we compare the results of the proposed NNIs from DAG_1 with the known
+// likelihoods of the actual NNIs already contained in DAG_2. This verifies we generate
+// the same result from adding NNIs to the DAG and updating as we do from using the
+// pre-NNI computation.
+TEST_CASE("Top-Pruning: Likelihoods with Proposed NNIs") {
+  // Build GPInstance with TPEngine and NNIEngine.
+  auto MakeTPEngine = [](const std::string& fasta_path, const std::string& newick_path,
+                         const std::string& tp_mmap_path,
+                         const std::string& gp_mmap_path) {
+    // Make GPInstance and TPEngine.
+    auto inst = GPInstanceOfFiles(fasta_path, newick_path, gp_mmap_path);
+    auto& dag = inst.GetDAG();
+    inst.MakeEngine();
+    GPEngine& gpengine = *inst.GetEngine();
+    auto edge_indexer = dag.BuildEdgeIndexer();
+    inst.MakeTPEngine(tp_mmap_path, true, true);
+    inst.MakeNNIEngine();
+    TPEngine& tpengine = inst.GetTPEngine();
+    tpengine.SetBranchLengths(gpengine.GetBranchLengths());
+    auto tree_collection = inst.GenerateCompleteRootedTreeCollection();
+    tpengine.SetChoiceMapByTakingFirst(tree_collection, edge_indexer);
+    tpengine.InitializeLikelihood();
+    tpengine.ComputeLikelihoods();
+    return inst;
+  };
+  // Build map from likelihood to edge in DAG.
+  auto BuildEdgeLikelihoodMap = [](GPInstance& inst) {
+    std::vector<RootedTree> tree_vector;
+    std::unordered_map<EdgeId, size_t> tree_id_map;
+    std::unordered_map<EdgeId, double> tp_likelihood_map;
+
+    auto& dag = inst.GetDAG();
+    auto& tpengine = inst.GetTPEngine();
+    auto tree_collection = inst.GenerateCompleteRootedTreeCollection();
+    const auto tree_source = tpengine.GetTreeSource();
+    // Populate tree vector and edge map.
+    for (const auto tree : tree_collection) {
+      tree_vector.push_back(tree);
+    }
+    for (EdgeId edge_id = 0; edge_id < dag.EdgeCountWithLeafSubsplits(); edge_id++) {
+      tree_id_map[edge_id] = tree_source[edge_id.value_];
+    }
+    // Compute likelihoods with TPEngine.
+    for (const auto& [edge_id, tree_id] : tree_id_map) {
+      std::ignore = tree_id;
+      auto likelihood = tpengine.GetTopTreeLikelihoodWithEdge(edge_id);
+      tp_likelihood_map[edge_id] = likelihood;
+    }
+    return tp_likelihood_map;
+  };
+  // Compare likelihoods from TPEngine. DAG_1 must be a subset of DAG_2.
```

Comment:
Let's have a comment explaining exactly what's being tested here-- it's that each likelihood in the one DAG matches at least one other likelihood in the second dag, right?

Reply:
How's this?
```
  // Compare likelihoods from TPEngine. Every edge in DAG_1 must yield a tree likelihood
  // shared by at least one edge in DAG_2.
  ```

Reply:
Updated.